pkg/controller/v1alpha2/applicationconfiguration/render.go

```diff
@@ -421,8 +419,154 @@ func addDataOutputsToDAG(dag *dependency.DAG, outs []v1alpha2.DataOutput, obj *u
+func (r *components) handleDependency(ctx context.Context, w *Workload, acc v1alpha2.ApplicationConfigurationComponent, dag *dag) ([]v1alpha2.UnstaifiedDependency, error) {
+	uds := make([]v1alpha2.UnstaifiedDependency, 0)
+
+	unsatsified, err := r.handleDataInput(ctx, acc.DataInputs, dag, w.Workload)
+	if err != nil {
+		return nil, errors.Wrapf(err, "handleDataInput for workload (%s/%s) failed", w.Workload.GetNamespace(), w.Workload.GetName())
+	}
+	if len(unsatsified) != 0 {
+		uds = append(uds, unsatsified...)
+		w.Unready = true
+	}
+
+	for i, ct := range acc.Traits {
+		trait := w.Traits[i]
+		unsatsified, err := r.handleDataInput(ctx, ct.DataInputs, dag, &trait.Object)
+		if err != nil {
+			return nil, errors.Wrapf(err, "handleDataInput for trait (%s/%s) failed", trait.Object.GetNamespace(), trait.Object.GetName())
+		}
+		if len(unsatsified) != 0 {
+			uds = append(uds, unsatsified...)
+			trait.Unready = true
+		}
+	}
+	return uds, nil
+}
+
+func makeUnsatisfiedDependency(obj *unstructured.Unstructured, s *dagSource, in v1alpha2.DataInput) v1alpha2.UnstaifiedDependency {
+	return v1alpha2.UnstaifiedDependency{
+		From: v1alpha2.DependencyFromObject{
+			TypedReference: runtimev1alpha1.TypedReference{
+				APIVersion: s.ObjectRef.APIVersion,
+				Kind:       s.ObjectRef.Kind,
+				Name:       s.ObjectRef.Name,
+			},
+			FieldPath: s.ObjectRef.FieldPath,
+		},
+		To: v1alpha2.DependencyToObject{
+			TypedReference: runtimev1alpha1.TypedReference{
+				APIVersion: obj.GetAPIVersion(),
+				Kind:       obj.GetKind(),
+				Name:       obj.GetName(),
+			},
+			FieldPaths: in.ToFieldPaths,
+		},
+	}
+}
+
+func (r *components) handleDataInput(ctx context.Context, ins []v1alpha2.DataInput, dag *dag, obj *unstructured.Unstructured) ([]v1alpha2.UnstaifiedDependency, error) {
+	uds := make([]v1alpha2.UnstaifiedDependency, 0)
 	for _, in := range ins {
-		dag.AddSink(in.ValueFrom.DataOutputName, obj, attaches, in.ToFieldPaths)
+		s, ok := dag.Sources[in.ValueFrom.DataOutputName]
+		if !ok {
+			return nil, errors.Wrapf(ErrDataOutputNotExist, "DataOutputName (%s)", in.ValueFrom.DataOutputName)
+		}
+		val, ready, err := r.checkSourceReady(ctx, s)
+		if err != nil {
+			return nil, errors.Wrap(err, "checkSourceReady failed")
+		}
+		if !ready {
+			uds = append(uds, makeUnsatisfiedDependency(obj, s, in))
+			return uds, nil
+		}
+
+		err = fillValue(obj, in.ToFieldPaths, val)
+		if err != nil {
+			return nil, errors.Wrap(err, "fillValue failed")
+		}
+	}
+	return uds, nil
+}
+
+func fillValue(obj *unstructured.Unstructured, fs []string, val string) error {
+	paved := fieldpath.Pave(obj.Object)
+	for _, fp := range fs {
+		err := paved.SetString(fp, val)
+		if err != nil {
+			return fmt.Errorf("paved.SetString() failed: %w", err)
+		}
+	}
+	return nil
+}
+
+func (r *components) checkSourceReady(ctx context.Context, s *dagSource) (string, bool, error) {
```

Comment:
I don't seem to see a Unit test here, maybe we can ask the student to add it?

Reply:
Let me answer all questions about the UT here.
All of the mentioned methods are covered in TestDependency. I have used `go tool cover` to check that.